<a href="https://colab.research.google.com/github/TimTree/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/module3-introduction-to-bayesian-inference/LS_DS_133_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 133

## Introduction to Bayesian Inference




## Assignment - Code it up!

Most of the above was pure math - now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up, and as a stretch goal - refactor your code into helpful reusable functions!

Specific goals/targets:

1. Write a function `def prob_drunk_given_positive(prob_drunk_prior, prob_positive, prob_positive_drunk)` that reproduces the example from lecture, and use it to calculate and visualize a range of situations
2. Explore `scipy.stats.bayes_mvs` - read its documentation, and experiment with it on data you've tested in other ways earlier this week
3. Create a visualization comparing the results of a Bayesian approach to a traditional/frequentist approach
4. In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/) - you could and should create something similar!

Stretch goals:

- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!

In [11]:
# TODO - code!

# There's a test that tests whether someone has taken a drug or not.
# If someone did take the drug, the test will say they took the drug 99% of the time.
# if someone did not take the drug, the test will say that they did not take the drug 99% of the time.
# 0.5% of all the people take the drug.
# If a random person tests positive, what's the probability they actually took the drug?

def prob_drunk_given_positive(prob_positive_drunk, prob_drunk_prior):
  return (prob_positive_drunk * prob_drunk_prior) / (prob_positive_drunk * prob_drunk_prior + 0.01 * (1 - prob_drunk_prior))

new_prior = prob_drunk_given_positive(0.99, 0.005)
new_prior2 = prob_drunk_given_positive(0.99, new_prior)
new_prior3 = prob_drunk_given_positive(0.99, new_prior2)
new_prior4 = prob_drunk_given_positive(0.99, new_prior3)
new_prior5 = prob_drunk_given_positive(0.99, new_prior4)
print(new_prior, new_prior2, new_prior3, new_prior4, new_prior5)

0.33221476510067116 0.9801000000000001 0.9997949506335922 0.9999979283737839 0.9999999790744397


In [12]:
new_prior = prob_drunk_given_positive(0.6, 0.005)
new_prior2 = prob_drunk_given_positive(0.6, new_prior)
new_prior3 = prob_drunk_given_positive(0.6, new_prior2)
new_prior4 = prob_drunk_given_positive(0.6, new_prior3)
new_prior5 = prob_drunk_given_positive(0.6, new_prior4)
print(new_prior, new_prior2, new_prior3, new_prior4, new_prior5)

0.23166023166023167 0.9476177941563569 0.9990795517093047 0.9999846452974989 0.9999997440844277


In [13]:
new_prior = prob_drunk_given_positive(0.999, 0.005)
new_prior2 = prob_drunk_given_positive(0.999, new_prior)
new_prior3 = prob_drunk_given_positive(0.999, new_prior2)
new_prior4 = prob_drunk_given_positive(0.999, new_prior3)
new_prior5 = prob_drunk_given_positive(0.999, new_prior4)
print(new_prior, new_prior2, new_prior3, new_prior4, new_prior5)

0.334225493476079 0.9804499651734304 0.9998004416354966 0.9999980020240521 0.9999999800002012


It's interesting that the probability of testing positive if drunk doesn't have to be high and can still lead to 99.9% certainty after just 3 attempts.

In [20]:
from scipy import stats
import numpy as np
data = [345,1,5,8,2,4645,32,6,8,5,346,7567563,354353,6,77,45,54,5,54]
stats.bayes_mvs(data,alpha=0.95)

(Mean(statistic=417240.0, minmax=(-418245.5145712617, 1252725.5145712616)),
 Variance(statistic=3380368906439.625, minmax=(1715576135879.5234, 6571202807527.828)),
 Std_dev(statistic=1810091.7762880102, minmax=(1309800.0366008254, 2563435.7428123346)))

In [21]:
def confidence_interval(data, confidence=.95):
  n = len(data)
  mean = sum(data)/n
  data = np.array(data)
  stderr = stats.sem(data)
  interval = stderr * stats.t.ppf((1 + confidence) / 2.0, n-1)
  return (mean , mean-interval, mean+interval)

confidence_interval(data)

(417240.0, -418245.5145712616, 1252725.5145712616)

The confidence intervals are the same between bayesian and frequentist.


Bayesian statistics makes inferences based on previous events. Frequentist statistics make interences based on one likelihood that something can happen.

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)